Redesigned algorithm based on DeFazio 2014

jolars opened the pull request:
This pull request is the result of several attempts to optimize the implementation of the algorithm. Most of this has been cleaned up now but originally involved benchmarking both the lighting-based implementation and the original from DeFazio. In the end, the latter turned out to perform best, although it has been augmented with bits and pieces from both the lighting and scikit-learn based implementation.

I have switched from Armadillo to Eigen, which lead to significant speed boosts and slight improvements in memory footprint. Here is a benchmark run similar to the one that I included with the proposal:

![image](https://user-images.githubusercontent.com/13087841/42291624-23c5e126-7fcd-11e8-9786-b0f45d8064cc.png)

I believe that this is close enough for most comparisons that additional optimizations in the future along with going parallel should put us ahead. One reason for why we might not reach the same speeds as in the cython-based implementations is that they use more aggressive optimization (-O3 compared to -O2, which seems to be default for R).

As you can see, however, there is still some way to go, and for computing the full path glmnet is still much faster in the L1-regularized case, probably due to the use of restricted sets via the strong rules. For L2-regression we are now very close performance-wise, however.

I have also included a brief vignette with these benchmarks. The data itself is produced from a script in data-raw. I am not sure that this is a good idea by the way, so if you have advice for a better way to store the data for the vignette I am all ears.

There is still only support for the binomial and gaussian families. I'll deal with that in the next pull request.

**Edit 6/7**
I should also note that the domain for binomial responses has been switched to {-1, 1}.

jolars committed:
Setup preliminary version of point-SAGA

jolars committed:
Only use key if number of coefficients is <= 10

jolars committed:
Test for ridge, lasso and elastic net penalties and increase tolerance for errors.

jolars committed:
Don't standardize for these tests.

jolars committed:
Rewrite algorithm after original SAGA article and add benchmarks vignette

* Switch from Armadillo to Eigen, which enables us to drop storage of nonzero indices and speeds up the algorithm.
* Translate DeFazios algorithm from the SAGA paper and replace the translation from scikit-learn, but keep some of the features such as convergence checks. Introduce fitting the intercept.
* Drop factory class for Prox and Family objects and create them directly.
* Add Link member for Family class.
* Switch from size_t to unsigned wherever integers are used.

jolars committed:
Setup a few tests for specifically testing the lambda path

These do not seem to work properly, or at least not compare directly to the values from glmnet. This should be checked in the future.

jolars committed:
Merge with target

Merge branch 'original-saga' of github.com:jolars/sgdnet into original-saga

# Conflicts:
#	src/Makevars
#	src/math.h
#	src/saga.h
#	src/sgdnet.cpp
#	src/utils.h

jolars committed:
Merge branch 'master' of github.com:jolars/sgdnet into original-saga

jolars committed:
Add missing s

jolars committed:
Reformat benchmark data, using raw data instead

jolars committed:
Drop leftovers from point-saga implementation

jolars committed:
Correct y axis label to non-normalized time

jolars committed:
Update binomial loss specification in documentation to match the internal representation

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -171,9 +171,9 @@ sgdnet.default <- function(x,
            # Transform response to {0, 1}, which is used internally
            y <- as.double(y)
-           y[y == min(y)] <- 0
+           y[y == min(y)] <- -1
```
Comment above this needs to be changed.

michaelweylandt commented on tests/testthat/test-lambda-path.R:
```diff
@@ -0,0 +1,43 @@
+  # has been implemented
+
+  for (sparse in c(TRUE, FALSE)) {
+    x <- Matrix::rsparsematrix(n, d, density = 0.2)
```
I like to have a `set.seed` call whenever I use random numbers in tests. Makes debugging much easier...

michaelweylandt commented on tests/testthat/test-gaussian.R:
```diff
@@ -20,11 +20,13 @@ test_that("gaussian regression with sparse and dense features work", {
-  fit_dense <- sgdnet(as.matrix(x), y, alpha = 0, intercept = FALSE)
+  fit_dense <- sgdnet(as.matrix(x), y, alpha = 0, intercept = FALSE,
+                      standardize = FALSE)
+  gfit <- sgdnet(x, y, alpha = 0, intercept = FALSE, standardize = FALSE)
```
Is this `gfit` being used?

michaelweylandt commented on src/sgdnet.cpp:
```diff
@@ -329,86 +137,81 @@ Rcpp::List SetupSgdnet(T&                   x,
-
   // Fit the path of penalty values
-  for (std::size_t penalty_ind = 0; penalty_ind < n_penalties; ++penalty_ind) {
+  for (unsigned penalty_ind = 0; penalty_ind < n_penalties; ++penalty_ind) {
```
Possibly think about changing this to something like `regularization_ind`. `penalty_ind` is definitely mathematically correct, but IME folks generally use "penalty" to refer to the base function and "lambda" to refer to the coefficient. E.g., in `argmin_{beta} |y - Xbeta|_2^2 + 3 |beta|_1` people would say it's L1-penalty, not a 3*L1 penalty.  (It's a distinction without a difference mathematically, but it's what people do...)

jolars replied:
You're right. I'll change it to lambda_ind and n_lambda.

michaelweylandt commented on src/sgdnet.cpp:
```diff
@@ -261,64 +73,60 @@ inline void AdaptiveTranspose(arma::mat& x) {
+  const bool     debug            = Rcpp::as<bool>(control["debug"]);
+
+  unsigned n_samples  = x.rows();
+  unsigned n_features = x.cols();
```
Isn't `unsigned` a type _modifier_? I'd expect `unsigned int` as the type here

jolars replied:
It is an alias for `unsigned int`. I am using simply to avoid verbosity but could change it if you wish.

michaelweylandt replied:
It looks like the standard guarantees `unsigned` is interpreted as `unsigned int` so don't worry about it: https://en.cppreference.com/w/cpp/language/types

michaelweylandt commented on data-raw/benchmarks.R:
```diff
@@ -0,0 +1,110 @@
+library(SparseM)
+library(Matrix)
+library(glmnet)
+library(gsoc18saga)
```
Add link to this since it's not a CRAN package.

michaelweylandt commented on src/families.h:
```diff
@@ -47,68 +45,65 @@ class Family {
     for (auto alpha_val : alpha_scaled) {
-      double L = L_scaling*(max_squared_sum + fit_intercept) + alpha_val;
+      double L =
+        L_scaling_*(max_squared_sum + static_cast<double>(fit_intercept))
```
Where is `fit_intercept` coming from here? It got dropped from the argument list.

jolars replied:
It is still there on line 41 I believe and is included in its called as this (in sgdnet.cpp): `family->StepSize(max_squared_sum, alpha, fit_intercept, n_samples);`

michaelweylandt replied:
Oops - my mistake. I didn't expand the "shrunk" part of the diff and thought this was the implementation of `PreprocessResponse` -- looks good.

michaelweylandt commented on src/families.h:
```diff
@@ -47,68 +45,65 @@ class Family {
+    double y_mu = Mean(y);
+    double y_sd = StandardDeviation(y);
+
+    if (y_sd == 0.0) y_sd = 1.0;
```
Does everything still work if `y` is constant? If it doesn't, then we just have `beta = 0` and `intercept = avg(y)` so we can special case this easily.

jolars replied:
We are using uncorrected standard deviation so are avoiding division by 0 in this case so this is not a problem (if I understood you correctly?).

michaelweylandt replied:
I was just wondering if everywhere else in the algorithm works fine with constant y. 

This would make a good (and easy) test case: if `y` is constant, then we should have `beta = 0`, `intercept = mean(y)` for the Gaussian. 

For the logistic, I think the intercept would diverge to `+/- infinity`, again with all `beta = 0`.

jolars replied:
Yeah, good thing you pointed it out. Computing the lambda path of course failed because lambda_max was initialized to 0. I have changed it so that a path of zeroes is returned in this case instead.

```r
x <- as.matrix(iris[, 1:4])
y <- rep(5, nrow(x))

sfit <- sgdnet(x, y)

coef(sfit)[, 1:10]
#> 5 x 10 sparse Matrix of class "dgCMatrix"
#>    [[ suppressing 10 column names ‘s0’, ‘s1’, ‘s2’ ... ]]
#>                                 
#> (Intercept)  1 1 1 1 1 1 1 1 1 1
#> Sepal.Length . . . . . . . . . .
#> Sepal.Width  . . . . . . . . . .
#> Petal.Length . . . . . . . . . .
#> Petal.Width  . . . . . . . . . .
```

I am currently not allowing constant y `family = "binomial"` because it results in a single class but I think you are right.

michaelweylandt commented on src/saga.h:
```diff
@@ -38,7 +38,7 @@
 // AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 // IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
-// ARE DISCLAIMED. IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE FOR
+// ARE DISCLAIMED. IN NO EVENT SHALL THE alphaENTS OR CONTRIBUTORS BE LIABLE FOR
```
Typo here

michaelweylandt commented on src/saga.h:
```diff
@@ -47,453 +47,436 @@
 }
 
-//' SAGA algorithm
+//' Lagged updates for L1-regularized regression
```
Did you mean to change this?

michaelweylandt commented:
Looks good - I didn't check the algorithmic details though. 

Which algorithm exactly is DeFazio 2014? The original SAGA paper or the finito paper?

jolars committed:
Correct domain in comment

jolars committed:
Set random seed in lambda tests

jolars committed:
Use fit from glmnet in sparse and dense gaussian tests

jolars committed:
Rename penalty_ind and n_penalties to lambda_ind and n_lambda respectively

We want to make it a little clear and avoid confusing the actual penalty withe its strength (lambda).

jolars commented:
I actually looked mostly at <https://github.com/adefazio/point-saga/blob/master/optimization/saga.pyx> but this is basically a carbon copy of the SAGA paper.

jolars committed:
Fix typo in license and update documentation for Saga()

jolars committed:
Remove n_lambda duplicate and update length before call

jolars committed:
Add utility function is_false() to avoid breaking R oldrel and keep using isFALSE functionality

michaelweylandt commented:
Ok - sounds good. (The logger at the end of the link prints "point-saga" instead of SAGA, but I think that's a typo)

jolars committed:
Fallback to all zero lambda path for constant response

jolars committed:
Test that a constant response results in a sparse solution with intercept at the mean

jolars committed:
Link to repo for gsoc18saga

jolars commented:
I'll merge this one in now. Thanks for the review!
